### PR TITLE
refactor(pay-form): decompose into smaller components

### PR DESCRIPTION
**Pull request by bolatovumar**
Address #2100

## Description:

Decomposed the pay form into smaller components and added unit tests and additional e2e tests.

## How Has This Been Tested?

Unit tests, e2e tests and manual testing.

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have reviewed and updated the documentation accordingly.
- [x] I have read the _CONTRIBUTING_ document.
- [x] I have added tests to cover my changes where needed.
- [x] All new and existing tests passed.
- [x] My commits have been squashed into a concise set of changes.

### CODE REVIEW

**Comment by korhaliv** [comment]
Hey @bolatovumar . Haven't looked at the code yet, but tests are failing. Could you please look into that first?

**Comment by mrfelton** [comment]
There was an issue with our testnet4 lnd node that could have been causing the failure here. I have restarted that node and I will rerun these tests on CI now

**Comment by bolatovumar** [comment]
@mrfelton thanks. Tests were passing on my end but I will re-run them again to confirm locally.

**Comment by bolatovumar** [comment]
@mrfelton I started going to down the path of using hooks and stuff but ran into some difficulties with how the react-spring library does hooks and since it's used quite a bit I left it as is.

You are definitely right about proptypes. I made sure proptypes that are defined are actually used but it's not obvious when you have a large list of them and they are being passed down. Would you rather have props passed down explicitly instead of doing something like `<Components {...props} />`?

**Comment by mrfelton** [comment]
> I started going to down the path of using hooks and stuff but ran into some difficulties with how the react-spring library does hooks and since it's used quite a bit I left it as is.

@bolatovumar Yes, that's fair enough. I don't want to complicate this PR with a mass conversion to hooks.

> I made sure proptypes that are defined are actually used but it's not obvious when you have a large list of them and they are being passed down. Would you rather have props passed down explicitly instead of doing something like <Components {...props} />?

We generally do this more explicitly since it makes the code a little easier to read being able to see what props are being passed into a component. The way you have it now you are passing all props to all child components (in `PayPanelBody` for example). It'd be better to explicitly pass through only the props that are needed for each child component so that there are no unused props being passed into a component.

A general note regarding prop passing - generally speaking, we are trying to refactor things such that we create multiple containers and connect components to the store as low as possible rather than connecting one single high level component and passing props down from there to everything else. Obviously that may not make sense in all situations.

**Comment by bolatovumar** [comment]
@mrfelton re:prop passing I did notice the pattern with multiple containers but also saw quite a few cases where the top level component accepted quite a few props and just passed them down to some child components and just went with that since most of the child components of the pay form are quite simple and have some tiny bit of top-level logic.

**Comment by korhaliv** [comment]
> @mrfelton re:prop passing I did notice the pattern with multiple containers but also saw quite a few cases where the top level component accepted quite a few props and just passed them down to some child components and just went with that since most of the child components of the pay form are quite simple and have some tiny bit of top-level logic.

Yes, there are instances of such a pattern in the current codebase. However, our goal is to get rid of that as much as possible. For reference see #2594 as one of the most recent moves in this direction.

**Comment by bolatovumar** [comment]
> > @mrfelton re:prop passing I did notice the pattern with multiple containers but also saw quite a few cases where the top level component accepted quite a few props and just passed them down to some child components and just went with that since most of the child components of the pay form are quite simple and have some tiny bit of top-level logic.
> 
> Yes, there are instances of such a pattern in the current codebase. However, our goal is to get rid of that as much as possible. For reference see #2594 as one of the most recent moves in this direction.

Thanks, I will take a look at that PR.

**Comment by bolatovumar** [comment]
Appveyor build seems to have failed for unrelated reasons.

File: renderer/components/Pay/PayAddressField.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      currentStep,
+      handlePayReqChange,
+      intl,
+      isLn,
+      network,
+      redirectPayReq,
+    } = this.props
+
+    const payReq_label =
+      currentStep === PAY_FORM_STEPS.address
```
Comment (korhaliv) [inline renderer/components/Pay/PayAddressField.js]: Can we re-write this in a more understandable way? These nested ternary operators look confusing

File: renderer/components/Pay/PayAmountFields.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+      initialAmountFiat,
+      isOnchain,
+      isQueryingFees,
+      lndTargetConfirmations,
+    } = this.props
+
+    const formState = formApi.getState()
+    const { isCoinSweep } = formState.values
+    const amountFieldsDisplayState =
+      currentStep === PAY_FORM_STEPS.amount
```
Comment (korhaliv) [inline renderer/components/Pay/PayAmountFields.js]: Same here

File: renderer/components/Pay/PayPanelHeader.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  return (
+    <>
+      <PayHeader
+        title={
+          <>
+            <FormattedMessage {...messages.send} /> {chainName} ({cryptoUnitName})
+          </>
+        }
+        type={isLn ? 'offchain' : isOnchain ? 'onchain' : null}
```
Comment (korhaliv) [inline renderer/components/Pay/PayPanelHeader.js]: and here

File: renderer/components/Pay/PaySummary.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+          address={payReq}
+          amount={amountInSats}
+          fee={getFee()}
+          isCoinSweep={isCoinSweep}
+          lndTargetConfirmations={lndTargetConfirmations}
+          mt={-3}
+          speed={speed}
+        />
+      )
+    } else if (isLn) {
```
Comment (korhaliv) [inline renderer/components/Pay/PaySummary.js]: else is not needed

File: renderer/components/Pay/PayPanelFooter.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  if (isLn && invoice) {
+    hasEnoughFunds = amountInSats <= channelBalance
+  } else if (isOnchain) {
+    hasEnoughFunds = amountInSats <= walletBalanceConfirmed
+  }
+
+  return hasEnoughFunds
+}
+
+const nextButtonText = (formState, props) => {
```
Comment (korhaliv) [inline renderer/components/Pay/PayPanelFooter.js]: can we rename this to `getNextButtonText` ? there is a local variable with the same name and in general `getX` is probably a better name for this kind of routine

File: renderer/components/Pay/PayAddressField.js (hunk 1 above)
Comment (bolatovumar) [reply renderer/components/Pay/PayAddressField.js]: will do

Comment (korhaliv) [reply renderer/components/Pay/PayAddressField.js]: also please rename `payReq_label ` to a camel case `payReqLabel`

**Comment by bolatovumar** [comment]
@korhaliv all your comments should be addressed now in the latest commit

**Pull request merged**